### PR TITLE
[5.8] Allow Carbon 2 and custom date handling

### PR DESCRIPTION
**Pull request by kylekatarnls**
This PR allow both Carbon `^1.26.3` and `^2.0` to work in Carbon and it allows the user to choose the class he want for dates he receive (such as `CarbonImmutable`, `Chronos` or simply `DateTime`) and/or intercept date generation to customize the date objects.

This PR refactor #25310 using facade instead of factory.

Some examples:
```php
Date::swap(\Carbon\CarbonImmutable::class);
// Assuming created_at is an Eloquent date property
var_dump(get_class($this->created_at)); // Carbon\CarbonImmutable
```
```php
// Apply French language on translatable methods for all generated dates
Date::swap(function (\Carbon\CarbonInterface $date) {
  return $date->locale('fr');
});
var_dump($this->created_at->locale); // fr
```
```php
Date::swap(new \Carbon\Factory([
  'locale' => 'en_AU',
  'timezone' => 'Australia/Sydney',
]));
var_dump($this->created_at->locale); // en_AU
var_dump($this->created_at->tzName); // Australia/Sydney
```

Note: optionally, the `Date` facade could be added to default config app.aliases.

### CODE REVIEW

File: src/Illuminate/Support/Facades/Date.php [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+use DateTimeInterface;
+use InvalidArgumentException;
+use Illuminate\Support\Carbon;
+
+/**
+ * @method static \Illuminate\Support\Carbon                         create($year = 0, $month = 1, $day = 1, $hour = 0, $minute = 0, $second = 0, $tz = null)                       Create a new Carbon instance from a specific date and time.
+ *                                                                                                                                                                                  If any of $year, $month or $day are set to null their now() values will
+ *                                                                                                                                                                                  be used.
+ *                                                                                                                                                                                  If $hour is null it will be set to its now() value and the default
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Facades/Date.php]: What's with all the extra whitespace?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/Facades/Date.php (hunk 1 above)
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: This is for alignment, it's get from auto-generated doc of the Carbon factory.

Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: Longest type is `\Symfony\Component\Translation\TranslatorInterface` so it pad whitespace on smaller types and longest method is `createSafe($year = null, $month = null, $day = null, $hour = null, $minute = null, $second = null, $tz = null)` so it pad whitespace on other methods.

Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: I see other faces does not describe methods and does not align anything, would you like the same for this Date facade?

File: src/Illuminate/Support/InteractsWithTime.php [hunk 2]
```diff
@@ -8,6 +8,7 @@
 trait InteractsWithTime
 {
     /**
+     * Get the number of seconds until the given DateTime.
```
Comment (mputkowski) [inline src/Illuminate/Support/InteractsWithTime.php]: I guess you added this by mistake...

Comment (kylekatarnls) [reply src/Illuminate/Support/InteractsWithTime.php]: Thanks! Will fix that.

File: src/Illuminate/Support/Facades/Date.php (hunk 1 above)
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: I know, I will compact it then put some links to the reference docs

Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: @GrahamCampbell done

File: src/Illuminate/Support/Facades/Date.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                $date = new static::$className($date->format('Y-m-d H:i:s.u'), $date->getTimezone());
+            }
+        }
+
+        if (static::$interceptor) {
+            return call_user_func(static::$interceptor, $date);
+        }
+
+        return $date;
+    }
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: What don't you understand here?

It's about giving different possibilities to the user:
- if Facade swap used, try to use it
- else:
  - try to call the method if it exists on the class (Carbon by default, but it can be changed with `use`)
  - else call it on Carbon original class then create the new object with it
    - we first try to create the new object using `instance`
    - if `instance` method does not exist, create it using constructor assuming the constructor is compatible with DateTime (matches the `if instance method or implements DateTimeInterface` in `use` method)
  - in both case apply the interceptor if specified by the user

There's nothing more.

**Comment by rny** [comment]
+1  Carbon 2 has important improvement because it provides toJson(), a standard way of transfer date in JSON.

File: src/Illuminate/Support/Facades/Date.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (mfn) [reply src/Illuminate/Support/Facades/Date.php]: What's the reason to have `static::$className` feature  *and* the fallback to `Carbon::` in `__callStatic`?

Why don't we go only for `static::$className`, why the fallback mechanism?

[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: Suppose you have `Date::use(SomeClass::class)`, then in the framework, Laravel calls `Date::someMethod()` if `someMethod` exists in `Carbon` but not in `SomeClass`, then we will do `Carbon::someMethod`, then create a `SomeClass` object from it. This way, inside Laravel we can continue to use `Date` as if it is the `Carbon` class no matter what other class would choose the end-user.

[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: Not agree. In my opinion, Laravel devs should benefit of all features of Carbon to create dates, and end-users should just have a way to convert those instances into different objects. But Laravel should not care about what class the user chose. When you have `Date::createFromTimestamp()` in the framework code (not the application code), it should not requires `createFromTimestamp` to be implemented in the class chosen by the end-user.

[the remark this answers is not in the document]
Comment (deleugpn) [reply src/Illuminate/Support/Facades/Date.php]: That's exactly where we disagree. If I configure the facade to not use Carbon, it should never use Carbon under no circumstances whatsoever.

[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: I'm not sure you get the feature, it's not a facade we will expose to the end-user. The user will not do `\Illuminate\Supports\Facades\Date::something()`, there is no need for the user to change his habits, if he used `now()` function or `\Illuminte\Support\Carbon`, he will just keep using them as before. If he want to use an other class, then he will do as before too `CarbonImmutable::now()` then if he run `Date::use(CarbonImmutable::class)`, he will get `CarbonImmutable` instances when calling `now()` too and when retrieving a date from the DB. That's the purpose. The only methods the end-user will need use from this facade is `use` or `intercept`.

Then, I'm not going to prevent Laravel developers to use Carbon internally, that would be dummy. And if some end-user want raw `DateTime` objects, of course it does not mean Laravel will use `DateTime` internally.

The `Date` facade is the common endpoint between the user desired class (could be any DateTimeInterface compatible class) and the dates created internally that can be used by the user and internally, dates will still be created with Carbon, no matter what the end-user want to work with.

[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: Just try it in a real app, run `Date::use(DateTime::class)` then get a date such as `$user->created_at`, you will understand why `$date = static::$className::$method(...$args);` is not enough. But if you find a easier solution that does not break **internal** usage of Carbon (and future usages) and still allow to custom the date type for the end-user, I'm open.

File: src/Illuminate/Support/Facades/Date.php [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            /** @var Carbon $date */
+            $date = Carbon::$method(...$args);
+            if (method_exists(static::$className, 'instance')) {
+                $date = static::$className::instance($date);
+            } else {
+                $date = new static::$className($date->format('Y-m-d H:i:s.u'), $date->getTimezone());
+            }
+        }
+
+        if (static::$interceptor) {
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: Yes, it just provide 2 ways to the user, one is easier but limited `use`, while `intercept` allow you to run any code to convert your instance.

[the remark this answers is not in the document]
Comment (mfn) [reply src/Illuminate/Support/Facades/Date.php]: IMHO there should be only clear way

[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: It must be simple for the user in priority. `Date::use(DateTime::class)` is simpler than `Date::intercept(function ($date) { return new DateTime($date->format('Y-m-d H:i:s.u'), $date->getTimezone()); });` but `Date::use` is not enough for custom conversions (change settings, creating the object in an other way). So an `$interceptor` and a `$className`, we can convert `$className` into `$interceptor` inside the `use` method, but it's just moving the code to an other place.

So what solution do you propose that would be as easy for the user we it comes to simple class swap and flexible enough for custom conversions?

[the remark this answers is not in the document]
Comment (deleugpn) [reply src/Illuminate/Support/Facades/Date.php]: `Date::swap(new MyOwnDateClass);` should be enough. If I want any interception, I can do in my class and the facade should not fallback to Carbon.

**Comment by kylekatarnls** [comment]
Code simplified according to @mfn suggestion to use only one method. Now all can be done with the `swap` method.

@GrahamCampbell An opinion on this?

File: src/Illuminate/Support/Facades/Date.php (hunk 1 above)
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: @GrahamCampbell Verbose documentation removed.

**Comment by kylekatarnls** [comment]
Hi @GrahamCampbell, your requested changes has been made. Any news on this?

**Comment by taylorotwell** [comment]
Can you explain the __callStatic method of the Date facade a bit? It looks a little complicated.

Also what is the `addRealSeconds` vs `addSeconds`?

**Comment by kylekatarnls** [comment]
`addSeconds` relies to native PHP DateTime and ignore DST while `addRealSeconds` use timestamps (ignore timezone). So in a case like cookie expiration: suppose you're in the Chicago timezone or any timezone having a DST. You create a cookie at midnight the 4 of november, and add 3 hours for expiration, in fact this cookie will expire 4 hours later due cloak change (https://www.timeanddate.com/time/change/usa/chicago), but if you use addRealSeconds/addRealMinutes/addRealHours, you get an actual 3 hours expiration.

So `addSeconds` can be useful in some cases, and most of the time, both will be equivalent (mostly if you work in UTC) but where I changed them `addRealSeconds` is more relevant. Those methods have been added in Carbon 1.25.0.

The `__callStatic` allows to pass various inputs to `swap`:
- class name, in this case we try to call the static method on it if it exists, else we call it on Carbon then cast the object to the asked class name.
- closure/callable, we call the static method on Carbon then pass the result to this callable and return what this callable returns.
- factory, this is the typical facade way.

If when `__callStatic` is called, there is no facade root yet selected, we set it to `Carbon::class`.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 5]
```diff
@@ -4,9 +4,11 @@
 
 use LogicException;
 use DateTimeInterface;
+use Carbon\CarbonInterface;
 use Illuminate\Support\Arr;
 use Illuminate\Support\Str;
 use Illuminate\Support\Carbon;
+use Illuminate\Support\Facades\Date;
```
Comment (mfn) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Will this work when using https://github.com/illuminate/database outside of the laravel framework?

I'm currently using illuminate/database, illuminate/events and illuminate/container outside of the laravel framework. The code where this is integrated does not use the Facades at all (though illuminate/support is a dependency which _contains_ the Facade declarations).

I'm asking because of all the components I've installed, except the illuminate/support definition, _none_ reference Facades in any way and I'm just worried this might break it.

Comment (kylekatarnls) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Carbon was loaded via illuminate/support, and Facades\Date is in the same illuminate/support so this is no different than before. Auto-load will do the job. And this facade has the advantage to use a fallback (Carbon) if it has no facade root defined. So it works with illumintate/support dependency only and you will even be able to set CarbonImmutable or other class for dates even using only some packages of illuminate.

I will try to make a database-only PoC.

**Comment by taylorotwell** [comment]
I still feel like I need a lot more explanation of the __callStatic method on Date Facade. Sorry! 😬 

Maybe I'm just slow at understanding it. Try adding a lot of comments to it at least for now so I can read through the logic and thinking a little better.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php (hunk 5 above)
Comment (mfn) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: > I will try to make a database-only PoC.

That would be highly appreciated!

The [usage instruction are up2date](https://github.com/illuminate/database#usage-instructions) and you can get things working pretty quick that way

Comment (kylekatarnls) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Tested requering illuminate/database only via composer (adding the modifications of this pull request), then running:
```php
echo get_class(User::first()->created_at)."\n";
Date::swap(DateTime::class);
echo get_class(User::first()->created_at)."\n";
Date::swap(\Carbon\CarbonImmutable::class);
echo get_class(User::first()->created_at)."\n";
```
Output:
```
Illuminate\Support\Carbon
DateTime
Carbon\CarbonImmutable
```
Here is the test project used.
https://github.com/kylekatarnls/illuminate-database-carbon-2-poc

Comment (mfn) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Thank you!

File: tests/Integration/Mail/SendingMailWithLocaleTest.php [hunk 6]
```diff
@@ -77,7 +77,7 @@ public function test_mail_is_sent_with_locale_updated_listeners_called()
 
         Mail::to('test@mail.com')->locale('es')->send(new TimestampTestMail);
 
-        $this->assertContains('nombre dentro de 1 día',
+        $this->assertRegExp('/nombre (en|dentro de) (un|1) día/',
```
Comment (mfn) [inline tests/Integration/Mail/SendingMailWithLocaleTest.php]: Why have these two tests changed in this PR?

Comment (kylekatarnls) [reply tests/Integration/Mail/SendingMailWithLocaleTest.php]: Why not? Carbon 2 is a major release, it has major changes such as translations harmonization with momentjs. It comes with output changes.

File: src/Illuminate/Support/Facades/Date.php [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        if (! $root) {
+            // Use Carbon class as root if not set
+            Date::swap($root = Carbon::class);
+        }
+
+        // If Date::swap(function () { ... }) with some closure or any callable has been called
+        if (is_callable($root)) {
+            // Then we create the date with Carbon then we pass it to this callable that can modify or replace
+            // This date object with any other value
+            return $root(Carbon::$method(...$args));
```
Comment (deleugpn) [inline src/Illuminate/Support/Facades/Date.php]: This feature appears to be inconsistent with Laravel facades. Calling `swap` should replace the object/class responsible for handling theat facade. This is an interceptor / adapter taking action by interpreting a special type of swap, which wasn't a real swap.
Perhaps if people want to intercept their date objects, they should wrap it in their own object / facade.
Real time facades are pretty good at that.

File: src/Illuminate/Support/Facades/Date.php [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+     * @throws \RuntimeException
+     */
+    public static function __callStatic($method, $args)
+    {
+        $root = null;
+
+        try {
+            $root = static::getFacadeRoot();
+        } catch (ReflectionException $exception) {
+            // We mute reflection exceptions as if have Carbon as fallback if the facade root is missing.
```
Comment (deleugpn) [inline src/Illuminate/Support/Facades/Date.php]: Following every other facade process, there would be one CarbonServiceProvider responsible for binding the accessor into the container, then we would never have a ReflectionException here.

File: src/Illuminate/Support/Facades/Date.php (hunk 7 above)
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: @deleugpn This provide a way far easier to convert date objects rather than creating an other class with again a __callStatic method to handle all possible static creations and swap to this class. I'm not in favor of a pure deletion of this.

File: src/Illuminate/Support/Facades/Date.php (hunk 8 above)
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: @deleugpn Good idea. I'm just wondering the impact of the provider dependency. First with this approach, if you use only some packages (illuminate/database only for example), then you have to run the `swap` manually to get dates properly created.

Then, right now, the service is optional (only handle default locale changes) and people may have chosen to disable the auto-discovery. These users will also have to call the `swap` before internal dates start to work.

**Comment by deleugpn** [comment]
Taylor's idea of comments was great for me to better see my problem with this. Since the beginning I see this deviating too much from the purpose of the Laravel Facades.
I guess now I understand the problem comes from the mixture of 2 classes into 1. 
The framework itself shouldn't use the facade because packages can be installed as standalone and that makes facades unavailable.
However, the framework needs to wrap the Carbon class in order to provide configuration of it for Laravel users.
The second thing is the facade, a facilitator for Laravel users to interact with the date object.
I guess if we were to split this up into a wrapper class (non-facade) and provide a facade for the wrapper, this would look much more like any other Laravel component.

**Comment by mfn** [comment]
> The framework itself shouldn't use the facade because packages can be installed as standalone and that makes facades unavailable.

Are you sure?

 I didn't test myself (and: I'm a user of illuminate/database standalone so this important for me) but @kylekatarnls gave me assurance, see https://github.com/laravel/framework/pull/25320#discussion_r219752474 .

File: src/Illuminate/Support/Facades/Date.php (hunk 8 above)
Comment (kylekatarnls) [reply src/Illuminate/Support/Facades/Date.php]: That's not true, facades works with only illuminate/support dependency, and the same dependency is needed for Carbon. But I see what you mean and will think of it.

**Comment by kylekatarnls** [comment]
@deleugpn is wrong on this particular point "that makes facades unavailable".

Facades are part of the illuminate/support, illuminate/database depends on it. So if you call swap on a facade then the facade can be used. The Date class proposed here allow not to call swap (as it fallback to Carbon).

@mfn I gave the code I used with no problem, you can test it yourself. As you can @deleugpn. I see your point about 2 classes to get a facade that looks more likely other ones as long as it can be done keeping esay tools for Laravel users configuration. I encourage you to test illuminate/support as stand-alone and swap and run facade. I think the solution should also allow this usage.

**Comment by deleugpn** [comment]
@mfn Yes, I'm sure. The code that @kylekatarnls prepared only works with Eloquent Models (which have their own __callStatic implementation) and the "Date Facade" that he provided here (which would give ReflectionException if he didn't silenced that).

Illuminate/Support only provides the Facade classes that relies on a container accessor. For that container accessor to be available, it needs to be bootstrapped and registered by the Service Providers. This is done in the Foundation Package. Laravel Zero (by Nuno Maduro) is a great example of how to use Facades outside of Laravel. You have to mimic the work of the Foundation package for that.

This "Date Facade" only works outside of Laravel because it's handling the situation where the Facades have not been bound into the container. This is why I think it should be split into 2 classes: a Wrapper and a Facade. The Wrapper is the thing that would be spread throughout the framework code itself. The facade is just a facilitator for Laravel users. The Wrapper could, then, provide a `DateWrapper::intercept()` in order to provide the extra functionalities without relying on a "hacky" `swap()`. The `Date::swap()` would allow Laravel users to change the DateWrapper into something else.

**Comment by kylekatarnls** [comment]
@deleugpn My understanding of the following code:
```php
static::$resolvedInstance[static::getFacadeAccessor()] = $instance;

if (isset(static::$app)) {
    static::$app->instance(static::getFacadeAccessor(), $instance);
}
```
Is that `static::$app` is not mandatory. So if I well understand this, a facade can be used with no app using only the `$resolvedInstance` array as singleton storage. And it seems not to me like an abuse, it seems like a well handled case. Then when you `resolveFacadeInstance` if the singleton exists it will not access to `$app` so it's not mandatory to bound it into the container.

And it's not Eloquent-related. The `ReflectionException` try-catch is not about facade mimic, it's only about making the `swap` call optional but we can do the same with a simple `isset(static::$resolvedInstance[$name])`

So I will still rely on the following system yet provided by Laravel Facades: use the app as facades container but fallback to the simple static array storage if missing.

**Comment by kylekatarnls** [comment]
According to the @deleugpn idea, I split the facade into 2 classes.

A real facade with no try-catch anymore and no __callStatic override. Just the default class fallback if not set in the `resolveFacadeInstance` method instead.

And a DateFactory that can handle callable, class name or factory as date handler.

We still can discuss the naming. But here I think we found a good middle-ground between complexity and ease of use.

**Comment by mfn** [comment]
🎉

**Comment by forestlanelabs** [comment]
@kylekatarnls, a bit confused by 2 classes - how do I completely replace Carbon with the standard DateTime? Don't want Carbon handling any dates and don't want to ever fallback to it.

**Comment by kylekatarnls** [comment]
First of all, pure datetime objects can be retrieve from any Carbon instance using `->toDateTime()`.

Then, in theory:
```php
Date::swap(\DateTime::class);
```

Would replace `Carbon` everywhere by `DateTime` internally.

But Laravel (and many libraries you may have installed) expect to have a `Carbon` compatible class. For example fresh Eloquent date in Laravel is created with `Date::now()` which will call `Carbon::now()` but if you switched to `DateTime` it calls `DateTime::now()` which does not exist.

So depending on you usage it actually may work.

But allowing `DateTime` would require some adjustment to make every internal features in Laravel compatible with it. And some Laravel third-party libraries you have in your project might simply not work with `DateTime`.

The purpose of `Date::swap()` is more to extend or switch to an equivalent class and note than `Carbon` allow you to call the native methods of `DateTime`, that means for most usage a function that expects a `DateTime` object will work like a charm receiving a `Carbon` instance instead. There is a difference in the way it's converted to JSON (`DateTime` output an object, `Carbon` a string), but you can easily align it with:
```php
Date::swap(new \Carbon\Factory([
    'toJsonFormat' => function ($date) {
        return $date->toDateTime();
    },
]));
```

So I would recommend to rely on inheritance using `DateTimeInterface` typing everywhere and the fact it's actually a `Carbon` instance or `DateTime` or any other class that could extend `DateTime` should make no difference. But when you really need to be sure you're not working with a `Carbon` object, you can:
```php
function purifyDate(\DateTimeInterface $date): \DateTimeInterface
{
    if ($date instance \Carbon\CarbonInterface) {
        return $date->toDateTime();
    }

    return $date;
}
```

**Comment by deleugpn** [comment]
@forestlanelabs I may be wrong here, I'm answering from my phone and only from memory. In theory, you can use Laravel's facade to swap the implementation to a brand new class of your choosing, which should probably adhere to php native DateTime object/interface. However, I think Laravel makes the conscious choice of binding itself to Carbon with Eloquent, so you'll have to work around date time features such as `$casts` and `timestamps`, I think. If you disable Eloquent Timestamp and never cast anything to date/datetime, you might get away with a lot by never landing on Carbon. Laravel will install Carbon anyway as a dependency, so it's a bit hard for you to handle that completely.
To summarise the idea behind two classes here: The facade is like any Laravel Facade: a statically accessible class that will resolve an implementation from the container for you and proxy any calls to it. The Date class that Laravel has now is a wrapper around Carbon to offer default Carbon configuration for Laravel users, as well as allowing swapping and/or reconfiguration.
